pkg/ovs/openflow/ofctrl_bridge.go

```diff
@@ -629,13 +673,80 @@ func (b *OFBridge) RetryInterval() time.Duration {
 	return b.retryInterval
 }
 
+func (b *OFBridge) queryTableFeatures() {
+	mpartRequest := &openflow13.MultipartRequest{
+		Header: openflow13.NewOfp13Header(),
+		Type:   openflow13.MultipartType_TableFeatures,
+		Flags:  0,
+	}
+	mpartRequest.Header.Type = openflow13.Type_MultiPartRequest
+	mpartRequest.Header.Length = mpartRequest.Len()
+	tableFeatureChan := make(chan *openflow13.MultipartReply)
+	b.mpartRplChs[mpartRequest.Header.Xid] = tableFeatureChan
+	go b.processTableFeatures(mpartRequest.Header.Xid, tableFeatureChan)
+	b.ofSwitch.Send(mpartRequest)
+}
+
+func (b *OFBridge) processTableFeatures(xid uint32, ch chan *openflow13.MultipartReply) {
+	header := openflow13.NewOfp13Header()
+	header.Type = openflow13.Type_MultiPartRequest
+	requests := make([]*openflow13.MultipartRequest, 0)
+	tableCount := 0
+	// Since the initial MultipartRequest doesn't specify any table ID, OVS will reply all tables' (except the hidden one)
+	// features in the reply. Although OpenFlow spec uses the Flags with a value 0 in the Multipart message to
+	// represent the message is complete, we can't ensure the last message sent from OVS is received as the last one.
+	// Here we count the number of tables received, and complete the loop when the number is the max value of OVS's tables.
+	for tableCount < int(LastTableID-1) {
+		select {
+		case rpl := <-ch:
+			request := &openflow13.MultipartRequest{
+				Header: header,
+				Type:   openflow13.MultipartType_TableFeatures,
+				Flags:  1,
+			}
+			// A MultipartReply message may have one or many OFPTableFeatures messages, and MultipartReply.Body is a
+			// slice of these messages.
+			for _, body := range rpl.Body {
+				request.Body = append(request.Body, body)
+			}
+			tableCount += len(rpl.Body)
+			request.Length = request.Len()
+			requests = append(requests, request)
+		}
+	}
+	close(ch)
```

Comment:
it's pretty unusual to have the _receiver_ close the channel

> we can't ensure the last message sent from OVS is received as the last one.

why, what would cause a re-ordering?

Reply:
I don't know the exact reason, but from my tests I observed this fact.

Reply:
this is a bit concerning IMO, because the whole scheme seems to rely on messages not being re-ordered (the last message sets a flag to indicated it's the last one). I kind of feel like we should try to understand why the re-ordering is happening. Are messages processed in parallel by different goroutines somewhere? Should all messages belonging to the same transaction be handled by the same goroutine?

Reply:
For the Multipart request/reply mesages we used to configure the table name,  it is processed in the same goroutine. I use the transactionID as the filter to ensure all the reply messages on the same request is sent to the same channel, and the consumer of the channel is running in one goroutine.

Reply:
Then why are the messages re-ordered? Some bug in ofnet?

Reply:
I don't think so. ofnet uses one goroutine to process all messages from the inbound stream (received from the unix domain socket). I used to test in ofnet, and printed the MultipartReply messages after ofnet Controller receives the message, the result also showed the order was not as expected. I am asking help from OVS team to see if it is about OVS design.

Reply:
With @ashish-varma help, we find the root cause is the goroutines in libOpenflow stream breaks the order. I have made a PR https://github.com/antrea-io/libOpenflow/pull/11 to fix libOpenflow issue. After that fix is merged, I will change this logic to use the Flags field to complete the loop. @antoninbas @tnqn